ocaml/lambda/translmod.ml

```diff
@@ -300,7 +300,8 @@ let init_shape id modl =
   let rec init_shape_mod subid loc env mty =
     match Mtype.scrape env mty with
       Mty_ident _
-    | Mty_alias _ ->
+    | Mty_alias _
+    | Mty_strengthen _ ->
```

Comment:
Note that we scrape the type before matching on it so we only get `Mty_strengthen` here if the type is abstract. This is exactly the same behaviour as before when we just dropped strengthening on abstract types, so this would be just `Mty_ident`. Your example doesn't trigger this code path at all since `S1` isn't abstract.

Reply:
Sorry - I really should have split the above into two separate comments.

1) I agree with your assessment that this is very like the old behavior and nothing is needed here.

2) Basically unrelatedly, I think the current interactions between explicit strengthening and recursive modules are a little weird.  I think we could allow the above example (?), but anyway it's fine to disallow it, at least until some horrifying use case for this kind of recursive module strengthening surfaces.  But if we do disallow it, the error needs a little work and should show up in the test suite.  It should at a minimum point to the actual recursive module reference, and in a perfect world would say explicitly that strengthening with a mutually defined module is disallowed.  (If this turns out to be hard, I think it's fine to ship with the bad error, but put it in the test suite with a CR)

Reply:
I fixed the error location and added a test. I also added a CR although I don't really agree that the error message is unclear or that this warrants a better error message. My understanding is that the general rule is that you can't refer to modules from the same recursive group in the module types. Why do you think that strengthening is special or different to, say, this?
```
module type S = sig module type T = sig end end

module rec M : N.T = struct end
and N : S = struct module type T = sig end end
```
As to allowing this, I'm not quite sure what, e.g., this would do:
```
module rec M : S with N = ... and N : S with M = ...
```

Reply:
Yes - you're right, this would be new territory and there's no reason to do it now.  Thanks for cleaning up the error.